Fix filename parser for new filenames

laurenmarietta opened the pull request:
Closes #311.

This PR brings `utils.filename_parser` up to snuff with the latest version of the DMS product filesystem.

In order to account for the weird guider files, I was able to get ahold of a TR (JWST-STScI-004800, SM-12) that defines the GUIDER data format:
<img width="400" alt="screen shot 2019-02-28 at 3 55 04 pm" src="https://user-images.githubusercontent.com/12650803/53601413-fe988200-3b79-11e9-9795-e442ece821a2.png">

Then, I found a [JDox entry](https://jwst-docs.stsci.edu/display/JDAT/Understanding+Associations?q=association%20candidate) explaining "Stage 2c" outlier detection associations products that have names of the form:
`jw<PPPPP><OOO><VVV>_<GGSAA>_<EEEEE>_<detector >_<AC_ID>_<suffix> .fits`

Between these two fixes, that seems to all the files in the new filesystem. I ended up not having to ask anyone in DMS about them, either. If I should alert someone that the stage 2c files aren't very well documented, though, I am happy to do so.

pep8speaks commented:
Hello @laurenmarietta, Thank you for submitting the Pull Request !

 - In the file [`jwql/utils/constants.py`](https://github.com/spacetelescope/jwql/blob/1177bae2f9150d596f8fdb0e7369ff9a21dd55cd/jwql/utils/constants.py), following are the PEP8 issues :

> [Line 50:22](https://github.com/spacetelescope/jwql/blob/1177bae2f9150d596f8fdb0e7369ff9a21dd55cd/jwql/utils/constants.py#L50): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 51:22](https://github.com/spacetelescope/jwql/blob/1177bae2f9150d596f8fdb0e7369ff9a21dd55cd/jwql/utils/constants.py#L51): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent

 - In the file [`jwql/utils/utils.py`](https://github.com/spacetelescope/jwql/blob/1177bae2f9150d596f8fdb0e7369ff9a21dd55cd/jwql/utils/utils.py), following are the PEP8 issues :

> [Line 79:101](https://github.com/spacetelescope/jwql/blob/1177bae2f9150d596f8fdb0e7369ff9a21dd55cd/jwql/utils/utils.py#L79): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (106 > 100 characters)
> [Line 93:5](https://github.com/spacetelescope/jwql/blob/1177bae2f9150d596f8fdb0e7369ff9a21dd55cd/jwql/utils/utils.py#L93): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (2)
> [Line 111:101](https://github.com/spacetelescope/jwql/blob/1177bae2f9150d596f8fdb0e7369ff9a21dd55cd/jwql/utils/utils.py#L111): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (102 > 100 characters)
> [Line 121:101](https://github.com/spacetelescope/jwql/blob/1177bae2f9150d596f8fdb0e7369ff9a21dd55cd/jwql/utils/utils.py#L121): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (104 > 100 characters)
> [Line 175:101](https://github.com/spacetelescope/jwql/blob/1177bae2f9150d596f8fdb0e7369ff9a21dd55cd/jwql/utils/utils.py#L175): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (103 > 100 characters)
> [Line 176:101](https://github.com/spacetelescope/jwql/blob/1177bae2f9150d596f8fdb0e7369ff9a21dd55cd/jwql/utils/utils.py#L176): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (145 > 100 characters)

If you have not done so, please consult the [`jwql` Style Guide](https://github.com/spacetelescope/jwql/blob/master/style_guide/style_guide.md)

Johannes-Sahlmann commented on jwql/utils/constants.py:
```diff
@@ -42,8 +42,19 @@
-FILE_SUFFIX_TYPES = ['uncal', 'cal', 'rateints', 'rate', 'trapsfilled',
-                     'i2d', 'x1d', 's2d', 's3d', 'dark', 'ami', 'crf']
+# Add guider suffixes (see JWST-STScI-004800, SM-12)
+FILE_SUFFIX_TYPES = ['stream', 'stacked_uncal', 'image_uncal', 'stacked_cal', 'image_cal']
```
Since you are separating into different suffix types anyways, I suggest to store these in separate variables, e.g. `GUIDER_SUFFIX_TYPES =  ['stream', 'stacked_uncal', 'image_uncal', 'stacked_cal', 'image_cal']` etc. and then construct FILE_SUFFIX_TYPES at the end
`FILE_SUFFIX_TYPES = GUIDER_SUFFIX_TYPES + ...`

I cannot yet thing about a use case, but it may even be useful to have the functionality of determining the suffix group based on a given type (i.e. I have a file suffix `amiavg` and have jwql tell me what this is), so one could think about a dictionary to store these lists.

laurenmarietta replied:
Good idea - done!

Johannes-Sahlmann commented on jwql/utils/utils.py:
```diff
@@ -75,45 +76,65 @@ def filename_parser(filename):
 
-    # Stage 3 filenames with target ID, e.g. "jw80600-o009_t001_miri_f1130w_i2d.fits"
+    # Stage 2c outlier detection filenames
+    # e.g. "jw94015002002_02108_00001_mirimage_o002_crf.fits"
```
I suggest to add all new examples to FILENAME_PARSER_TEST_DATA in `test_utils.py`.

laurenmarietta replied:
Oh, great point! Sorry I forgot this on the first go.

bourque commented:
@laurenmarietta Thanks for doing all of that research and getting to the bottom of all of it!  

FYI, I pushed a commit where I tried to consolidate all of the various documents that describe JWST filename conventions and stuck it in the `References` section of the module docstring.

I guess the frustrating thing about this is ultimately all of these filename conventions are documented _somewhere_, but there doesn't seem to exist one document that lists and describes every possibility -- that would be nice to have!  But I suppose the creation and maintenance of such a document is beyond our control.  At least `filename_parser()` will serve as a decent reference itself.

laurenmarietta commented:
I was actually thinking the same thing about how long the function is, but didn't think it necessarily needed to be moved, since there are only 3 other small functions currently in `utils.py` right now.

If you both think it would be better to put it in its own function, though, I can move it in this PR.

laurenmarietta commented:
Thanks to Johannes' reminder to add tests, I realized that the stage 2c data actually were being parsed by the stage 1 & 2 regex. I added a fix for that ,and also added a new `filename_type_name` key to the output of `filename_parser`, so we now can tell kind of filename was parsed.

bourque commented:
@laurenmarietta To reply to your earlier comment --Ok, we can keep `filename_parser()` in `utils.utils` for now.  If it starts to become burdensome then we can perhaps move things around in a future PR.

bourque commented on jwql/utils/utils.py:
```diff
@@ -126,24 +154,59 @@ def filename_parser(filename):
     # Raise error if unable to parse the filename
     try:
         filename_dict = jwst_file.groupdict()
+        filename_dict['filename_type'] = name_match
```
@laurenmarietta Great idea!

Johannes-Sahlmann commented on jwql/tests/test_utils.py:
```diff
@@ -32,6 +32,7 @@
     {'activity': '02',
```
I apologize in advance of what's coming:
pylint report for this file: `Your code has been rated at -9.31/10`, mostly because of identation.
(pycodestyle also complains about it).
I leave it up to you to fix or not, I do not mind leaving it as is.

laurenmarietta replied:
Gah! PyCharm is great for a lot of things, but its linter lets me get away with a lot. I'll try to clean it up a bit.

Johannes-Sahlmann commented on jwql/utils/constants.py:
```diff
@@ -42,8 +42,22 @@
+GUIDER_SUFFIX_TYPES = ['stream', 'stacked_uncal', 'image_uncal', 'stacked_cal', 'image_cal']
+
+GENERIC_SUFFIX_TYPES = ['uncal', 'cal', 'rateints', 'rate', 'trapsfilled', 'i2d',
+                       'x1dints', 'x1d', 's2d', 's3d', 'dark', 'crfints',
```
pylint/pydocstyle complain about indentation here too.
